### PR TITLE
Bind compatibility GDExtension methods removed in #88418

### PR DESCRIPTION
**Pull request by lyuma**
- Reproducible in 4.3dev6
- System Information: Godot v4.3.dev (6118592c6) - Windows 10.0.19045 - Vulkan (Forward+) - dedicated NVIDIA GeForce RTX 3090 (NVIDIA; 31.0.15.3713) - Intel(R) Core(TM) i9-9900K CPU @ 3.60GHz (16 Threads)

These functions were likely not used, but we must ensure they are still bound to ensure API stability.

Without this PR, rust addons give this error:
```
ERROR: Panic msg:
  Failed to load class method GDExtension::open_library (hash 852856452).
  Make sure gdext and Godot are compatible: https://godot-rust.github.io/book/gdext/advanced/compatibility.html
```

`GDExtension::open_library()` with hash 852856452 needs to be re-added as a bind_compatibility_method.

### CODE REVIEW

**Comment by dsnopek** [comment]
We removed these methods without providing compatibility in PR https://github.com/godotengine/godot/pull/88418, because there is no way to use them correctly.

If we do need to add them back for compatibility, then we should make them do nothing (and in the case of `open_library()` return an error). These methods never should have been bound, and using them will lead to problems.

**Comment by Bromeon** [comment]
In the godot-rust bindings, I explicitly [removed these 3 functions](https://github.com/godot-rust/gdext/blob/cb41673a0ab802145543a8f371abfcb1f0be10f9/godot-codegen/src/special_cases/special_cases.rs#L49-L52). That way, even when using it with older API levels (4.1 or 4.2), it's not possible to accidentally use them.

**Comment by lyuma** [comment]
Right, but I have gdextension dlls that are probably not using the latest version of the bindings and work fine on Godot 4.2. Even if we remove them from the latest rust bindings, this is still technically a compat break since dlls have to be rebuilt.

File: core/extension/gdextension.cpp [hunk 1]
```diff
@@ -35,6 +35,7 @@
 #include "core/object/method_bind.h"
 #include "core/os/os.h"
 #include "core/version.h"
+#include "gdextension.compat.inc"
```
Comment (AThousandShips) [inline core/extension/gdextension.cpp:38]: Should be moved above, making it:
```cpp
#include "gdextension.h"
#include "gdextension.compat.inc"

#include "core/config/project_settings.h"
```

File: core/extension/gdextension.h [hunk 2]
```diff
@@ -136,6 +136,14 @@ class GDExtension : public Resource {
 		INITIALIZATION_LEVEL_EDITOR = GDEXTENSION_INITIALIZATION_EDITOR
 	};
 
+protected:
+#ifndef DISABLE_DEPRECATED
+	Error _open_library_bind_compat_88418(const String &p_path, const String &p_entry_symbol);
+	void _close_library_bind_compat_88418();
+	void _initialize_library_bind_compat_88418(InitializationLevel p_level);
+	static void _bind_compatibility_methods();
+#endif
```
Comment (AThousandShips) [inline core/extension/gdextension.h:145]:
```suggestion
#endif

```

File: core/extension/gdextension.cpp (hunk 1 above)
Comment (lyuma) [reply core/extension/gdextension.cpp:38]: When I tried doing this I got an error from style checks:
https://github.com/godotengine/godot/actions/runs/8935217873/job/24543349469

Do you know what I did wrong?

File: core/extension/gdextension.h (hunk 2 above)
Comment (lyuma) [reply core/extension/gdextension.h:145]: Same, the style checks [forced me to remove the blank line after #endif](https://github.com/godotengine/godot/actions/runs/8935217873/job/24543349469)

File: core/extension/gdextension.cpp (hunk 1 above)
Comment (AThousandShips) [reply core/extension/gdextension.cpp:38]: You missed the empty line, the header should always be separate

File: core/extension/gdextension.h (hunk 2 above)
Comment (AThousandShips) [reply core/extension/gdextension.h:145]: I don't see that in the ci

File: core/extension/gdextension.cpp (hunk 1 above)
Comment (lyuma) [reply core/extension/gdextension.cpp:38]: aha, I didn't know the empty line mattered

File: core/extension/gdextension.h (hunk 2 above)
Comment (lyuma) [reply core/extension/gdextension.h:145]: Ah I meant to link this one:
https://github.com/godotengine/godot/actions/runs/8935205337/job/24543315445

I wonder if the issue is I put the blank line after public: instead of between #endif and public:

Comment (AThousandShips) [reply core/extension/gdextension.h:145]: Yes that's not the same as I suggested, normally there should be one between any code and any access specifier, might not work for `endif` though

**Comment by Bromeon** [comment]
> Right, but I have gdextension dlls that are probably not using the latest version of the bindings and work fine on Godot 4.2. Even if we remove them from the latest rust bindings, this is still technically a compat break since dlls have to be rebuilt.

As I wrote on RocketChat, I think there are some misunderstandings.

First, what we try to achieve with compatibility on godot-rust side is that you can develop against API level 4.1 and still use the extension in Godot 4.2 and 4.3. **This is the case on latest `master`.**

If we encounter bugs, it can unfortunately happen that binary compatibility is affected. Not just the example here with 3 `GDExtension` methods, but in some cases we fixed soundness issues regarding how data was passed over FFI. So there are benefits to remaining up-to-date either way. Especially since we [spend a lot of effort in maintaining compatibility](https://godot-rust.github.io/book/toolchain/compatibility.html#current-guarantees) with older Godot versions -- you can even still use 4.0 if you like -- so you're not missing out on compatibility when updating.

The second and much more important point: godot-rust has supported lazy loading of function pointers [for more than half a year](https://github.com/godot-rust/gdext/pull/458). So it explicitly offers different trade-offs depending on the extension's needs. This is [documented](https://godot-rust.github.io/docs/gdext/master/godot/#cargo-features) on the main API docs page.

TLDR: like all software, godot-rust is improved over time and it is advised to remain reasonably up-to-date. We can't be responsible for bugs fixed in the past because recompiling DLL is a problem. Furthermore, you can enable lazy loading if that fits your workflow better.

**Comment by Bromeon** [comment]
Note that my above comment merely clarifies the situation around godot-rust.

The core discussion here -- breaking API changes in the presence of severe bugs -- can be quite nuanced.
Some thoughts:
- A "no breaking changes ever" policy would need alternative ways from discouraging the usage of certain functions -- documentation, error messages, etc.
- Such a policy can actively hide incorrect usage of the GDExtension API. If the usage no longer compiles, it forces the user to correct things, instead of possibly worse alternatives (logic errors or even UB).
- Obviously, removal should never be done lightly. Certainly not in "we now have a better API for this" cases. Imo only in "there's no way to use the previous correctly, at all".

File: core/extension/gdextension.h [hunk 3]
```diff
@@ -136,6 +136,15 @@ class GDExtension : public Resource {
 		INITIALIZATION_LEVEL_EDITOR = GDEXTENSION_INITIALIZATION_EDITOR
 	};
 
+protected:
+#ifndef DISABLE_DEPRECATED
+	Error _open_library_bind_compat_88418(const String &p_path, const String &p_entry_symbol);
```
Comment (AThousandShips) [inline core/extension/gdextension.h:141]:
```suggestion
	Error _open_library_bind_compat_91502(const String &p_path, const String &p_entry_symbol);
```
Should be named for the PR that added them to make it easier to track

**Comment by dsnopek** [comment]
Discussed at the GDExtension meeting, and even though we decided not add compatibility methods for these, this PR is good and there would be no harm in merging this. However, this isn't a policy change in favor of never breaking compatibility ever.

**Review by dsnopek — APPROVED** [review]

**Review by AThousandShips — APPROVED** [review]

File: core/extension/gdextension.h [hunk 4]
```diff
@@ -136,6 +136,15 @@ class GDExtension : public Resource {
...
+	void _close_library_bind_compat_88418();
+	void _initialize_library_bind_compat_88418(InitializationLevel p_level);
```
[the remark this answers is not in the document]
Comment (akien-mga) [reply core/extension/gdextension.h:143]: I think it's fine to use the number of the PR that removes these APIs, and should therefore have added the compat bindings. This PR is just a fixup to what was missing in #88418.

[the remark this answers is not in the document]
Comment (AThousandShips) [reply core/extension/gdextension.h:143]: Granted, but it might make it harder to find the PR itself

**Comment by akien-mga** [comment]
Thanks!
